vortex-datafusion/src/lib.rs

```diff
@@ -115,87 +167,250 @@ impl TableProvider for VortexInMemoryTableProvider {
 
     fn supports_filters_pushdown(
         &self,
         filters: &[&Expr],
     ) -> DFResult<Vec<TableProviderFilterPushDown>> {
-        // TODO(aduffy): add support for filter pushdown
-        Ok(filters
+        // In the case the caller has configured this provider with filter pushdown disabled,
+        // do not attempt to apply any filters at scan time.
+        if self.options.disable_pushdown {
+            return Ok(filters
+                .iter()
+                .map(|_| TableProviderFilterPushDown::Unsupported)
+                .collect());
+        }
+
+        // Verify for each filter whether its expression tree consists solely of operations we know
+        // how to pushdown, and that they only reference columns in our source.
+        // TODO(aduffy): figure out if we actually need to do this, or what guarantees about the
+        //  filters that DataFusion provides.
+        let schema_columns: HashSet<String> = self
+            .schema_ref
+            .fields
+            .iter()
+            .map(|field| field.name().clone())
+            .collect();
+
+        filters
             .iter()
-            .map(|_| TableProviderFilterPushDown::Unsupported)
-            .collect())
+            .map(|expr| {
+                if can_be_pushed_down(expr, &schema_columns)? {
+                    Ok(TableProviderFilterPushDown::Exact)
+                } else {
+                    Ok(TableProviderFilterPushDown::Unsupported)
+                }
+            })
+            .try_collect()
     }
 }
 
+/// Construct an operator plan that executes in two stages.
+///
+/// The first plan stage only materializes the columns related to the provided set of filter
+/// expressions. It evaluates the filters into a row selection.
+///
+/// The second stage receives the row selection above and dispatches a `take` on the remaining
+/// columns.
+fn make_filter_then_take_plan(
+    schema: SchemaRef,
+    filter_exprs: Vec<Expr>,
+    filter_projection: Vec<usize>,
+    array: Array,
+    output_projection: Vec<usize>,
+) -> Arc<dyn ExecutionPlan> {
+    let struct_array = StructArray::try_from(array).unwrap();
+
+    let filter_struct = struct_array
+        .project(filter_projection.as_slice())
+        .expect("projecting filter struct");
+
+    let row_selector_op = Arc::new(RowSelectorExec::new(&filter_exprs, &filter_struct));
+
+    Arc::new(TakeRowsExec::new(
+        schema.clone(),
+        &output_projection,
+        row_selector_op.clone(),
+        &struct_array,
+    ))
+}
+
+/// Check if the given expression tree can be pushed down into the scan.
+fn can_be_pushed_down(expr: &Expr, schema_columns: &HashSet<String>) -> DFResult<bool> {
+    // If the filter references a column not known to our schema, we reject the filter for pushdown.
+    // TODO(aduffy): is this necessary? Under what conditions would this happen?
+    let column_refs = get_column_references(expr);
+    if !column_refs.is_subset(schema_columns) {
```

Comment:
not sure how datafusion produces the output schema but if you have `SELECT A WHERE A > B` depending on the internal implementation it might not have `B` in the output schema

Reply:
DF differentiates between the table schema and the output projection, and filters can reference things that are not in the output projection but are in the table schema. They explain this behavior in https://docs.rs/datafusion/latest/datafusion/datasource/provider/trait.TableProvider.html#tymethod.scan under the `Filters` section.

The thing I was more concerned about is I don't know if it can pushdown filters that reference columns in other tables, like broadcast joins. Maybe this is too defensive